Interpret newlines in tooltips

wentasah opened the pull request:
This commit adds support for newlines in tooltips. The implemented behavior matches the behavior of Graphviz SVG output, but is not in-line with the documentation (i.e. Graphviz documentation differs from implementation).

According to the documentation [1], tooltips are only supported for svg and cmap outputs and not xdot (here xdot.py deviates from the documentation, but for good reasons) and the value should be escString. escString should support newlines only for label, headlabel or taillabel attributes, not for tooltips [2]. Nevertheless, the svg output supports them. But, contrary to the documentation, SVG tooltips do not support capital letter escape sequences like \N, \G and \E. Only \n, \l, \r and \\ seem to be supported. The first three are all interpreted as a new line, different alignments specified by the documentation are not distinguished. The same is implemented in this commit.

[1]: https://graphviz.org/docs/attrs/tooltip/
[2]: https://graphviz.org/docs/attr-types/escString/

Before this PR, my graph looks like this:
![xdot-orig](https://github.com/user-attachments/assets/12bec46b-1bf9-4e4c-b932-8a0d01883354)

With this PR, it looks this way:
![xdot-fixed](https://github.com/user-attachments/assets/e0e066d5-9367-4500-9236-ff087d284ee9)

jrfonseca commented:
You're right.

However this logic should be moved into the parser, not the UI.

I also noticed that the charset for these attributes is not always utf-8, because these are original dot attributes, and not the xdot's `_xxx_` generated attributes which are guaranteed to be in utf-8.

So please rebase on top of 0b91d2eea87f41b68cd708e2c5d647c1026fe957 and move your code to:

```diff
diff --git a/xdot/ui/_xdotparser.py b/xdot/ui/_xdotparser.py
index c9c721a..e88f57b 100644
--- a/xdot/ui/_xdotparser.py
+++ b/xdot/ui/_xdotparser.py
@@ -357,6 +357,7 @@ class XDotParser(DotParser):
                 shapes.extend(parser.parse())
         url = self.decode_attr(attrs, 'URL')
         tooltip = self.decode_attr(attrs, 'tooltip')
+        # FIXME call interpret_esc_nl here
         node = elements.Node(id, x, y, w, h, shapes, url, tooltip)
         self.node_by_name[id] = node
         if shapes:
@@ -378,6 +379,7 @@ class XDotParser(DotParser):
             src = self.node_by_name[src_id]
             dst = self.node_by_name[dst_id]
             tooltip = self.decode_attr(attrs, 'tooltip')
+            # FIXME call interpret_esc_nl here
             self.edges.append(elements.Edge(src, dst, points, shapes, tooltip))
 
     def parse(self):
```

wentasah commented:
Rebased and updated. It works on my graphs and I also run the test suite locally. I didn't observe anything suspicious in test output.

And thanks for pointing to the place in the parser. I thought the parser would be a better place, but I wasn't able to figure out the proper location in the code (I was looking in `dot/parser.py`).

jrfonseca commented:
Thank you!
